request.go

```diff
@@ -55,7 +59,44 @@ func requestFromPacket(pkt hasPath) *Request {
 
 // NewRequest creates a new Request object.
 func NewRequest(method, path string) *Request {
-	return &Request{Method: method, Filepath: cleanPath(path)}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Request{Method: method, Filepath: cleanPath(path),
+		ctx: ctx, cancelCtx: cancel}
+}
+
+// Context returns the request's context. To change the context,
+// use WithContext.
+//
+// The returned context is always non-nil; it defaults to the
+// background context.
+//
+// For incoming server requests, the context is canceled when the
+// request is complete or the client's connection closes.
+func (r *Request) Context() context.Context {
+	if r.ctx != nil {
+		return r.ctx
+	}
+	return context.Background()
+}
+
+// WithContext returns a copy of r with its context changed to ctx.
+// The provided ctx must be non-nil.
+func (r *Request) WithContext(ctx context.Context) *Request {
+	if ctx == nil {
+		panic("nil context")
+	}
+	r.stateLock.Lock()
+	defer r.stateLock.Unlock()
+	r2 := &Request{
```

Comment:
This sort of explicit copying of each field is error-prone, since if one adds a new field it would need to be added here. This is why the http.Request.WithContext method uses `*r2 = *r` shallow copy. But in our case, we have both: a Mutex that needs to not be copied, and a cancelFunc that we would need to clear (with the complications of “how to ensure it still gets called if this context is replaced”.)

Reply:
I have this changed already locally. After reading the http request/context code I thought that method was much cleaner. I have a couple queued up commits that clean up the lock usage a bit, convert the lock to a pointer (which you can copy then replace)  and uses the *r2 = *r for copying.

Reply:
Oh. And the cancel would be set to nil post copy. The original request would still be tracked and still be closed in the server code. Calling WithContext doesn't replace the one in memory (scoped in packetworker) nor does it replace the one tracked in RequestServer.openRequests. Those would still be around and get closed properly.

Reply:
👍